src/paperless_tesseract/parsers.py

```diff
@@ -45,13 +45,22 @@ def get_thumbnail(self):
         """
 
         out_path = os.path.join(self.tempdir, "convert.png")
+        gs_out_path = os.path.join(self.tempdir, "gs_out.png")
 
         # Run convert to get a decent thumbnail
+
+        # https://github.com/danielquinn/paperless/issues/447
+        # call gs first
+        cmd = ["gs", "-q", "-sDEVICE=pngalpha",
```

Comment:
Should we only use this as a fallback? I.e. should we try to execute ghostscript and fall back to the "old way" if it doesn't exist?

I was thinking about the function being something like this:

```python
    def get_thumbnail(self):
        """
        The thumbnail of a PDF is just a 500px wide image of the first page.
        """

        out_path = os.path.join(self.tempdir, "convert.png")
        gs_out_path = os.path.join(self.tempdir, "gs_out.png")

        # Extract the first PDF page as a PNG using Ghostscript, if available, to use as a thumbnail.
        # https://github.com/danielquinn/paperless/issues/447
        try:
            cmd = [
                self.GHOSTSCRIPT,
                "-q",
                "-sDEVICE=pngalpha",
                "-o", gs_out_path,
                self.document_path
            ]
            if not subprocess.Popen(cmd).wait() == 0:
                raise ParseError("Thumbnail (gs) failed at {}".format(cmd))
            thumbnail_source_path = gs_out_path
        except FileNotFoundError:
            thumbnail_source_path = "{}[0]".format(self.document_path)

        # Run convert to get a decent thumbnail
        run_convert(
            self.CONVERT,
            "-scale", "500x5000",
            "-alpha", "remove",
            thumbnail_source_path,
            out_path
        )

        return out_path
```

Or, as a diff:

```diff,python
     def get_thumbnail(self):
         """
         The thumbnail of a PDF is just a 500px wide image of the first page.
         """
 
         out_path = os.path.join(self.tempdir, "convert.png")
         gs_out_path = os.path.join(self.tempdir, "gs_out.png")
 
-        # Run convert to get a decent thumbnail
-
+        # Extract the first PDF page as a PNG using Ghostscript, if available, to use as a thumbnail.
         # https://github.com/danielquinn/paperless/issues/447
-        # call gs first
-        cmd = ["gs", "-q", "-sDEVICE=pngalpha",
-               "-o", gs_out_path, self.document_path]
-        if not subprocess.Popen(cmd).wait() == 0:
-            raise ParseError("Thumbnail (gs) failed at {}".format(cmd))
-        # then run convert on the output from gs
+        try:
+            cmd = [
+                self.GHOSTSCRIPT,
+                "-q",
+                "-sDEVICE=pngalpha",
+                "-o", gs_out_path,
+                self.document_path
+            ]
+            if not subprocess.Popen(cmd).wait() == 0:
+                raise ParseError("Thumbnail (gs) failed at {}".format(cmd))
+            thumbnail_source_path = gs_out_path
+        except FileNotFoundError:
+            thumbnail_source_path = "{}[0]".format(self.document_path)
+
+        # Run convert to get a decent thumbnail
         run_convert(
             self.CONVERT,
             "-scale", "500x5000",
             "-alpha", "remove",
-            "gs_out_path",
+            thumbnail_source_path,
             out_path
         )
 
         return out_path
```

Reply:
I like the idea of a fallback. But in this case if ghostscript is not available we depend on imagemagick, which may or may not fail on parsing.

What if we make gs a requirement and make sure that it is available by adding it to the binaries check in [checks.py](https://github.com/the-paperless-project/paperless/blob/6d8e4ad6bf92094149e2335a8f30cb976475fe70/src/paperless/checks.py#L70)?

Reply:
Because ImageMagick is using `Ghostscript` internally for its PDF operations, Ghostscript is already a strict dependency of Paperless (although not explicitly checked in `checky.py`).
So that `try: ... except:` is unneeded.
Actually, we should try the fast single pass `"{}[0]"` approach first, and then fall back to the slower `Ghostscript + ImageMagick` workaround.

Reply:
Resolved by https://github.com/the-paperless-project/paperless/pull/494/commits/448e5c201bbf4d195ea657d519d71fca34222e08